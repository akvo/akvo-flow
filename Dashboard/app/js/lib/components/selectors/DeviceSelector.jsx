import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line import/no-unresolved
import Checkbox from 'akvo-flow/components/Checkbox';

export default class DeviceSelector extends React.Component {
  state = {
    isAccordionOpen: false,
  }

  onAccordionClick = () => {
    const { isAccordionOpen } = this.state;
    this.setState({ isAccordionOpen: !isAccordionOpen });
  }

  render() {
    const { isAccordionOpen } = this.state;
    const { deviceGroups, deviceGroupNames, onCheck } = this.props;

    const accordionClass = `accordion ${isAccordionOpen && 'active'}`;
    const panelStyle = isAccordionOpen ? { display: 'block' } : { display: 'none' };

    return (
      <div>
        {Object.keys(deviceGroups).map(dgId => (
          <div key={dgId}>
            <div
              className={accordionClass}
              onClick={this.onAccordionClick}
              onKeyPress={this.onAccordionClick}
            >
              {deviceGroupNames[dgId]}
            </div>

            <div className="panel" style={panelStyle}>
              {Object.keys(deviceGroups[dgId]).map(deviceId => (
                <div key={deviceId}>
                  <Checkbox
                    id={deviceId}
                    name={deviceId}
                    checked={deviceGroups[dgId][deviceId].checked}
                    onChange={onCheck}
                  />

                  <label id={deviceId} htmlFor={deviceId}>
                    {deviceGroups[dgId][deviceId].name}
                  </label>
                </div>
              ))}
            </div>
          </div>
        ))}
      </div>
    );
  }
}

DeviceSelector.propTypes = {
  deviceGroups: PropTypes.object.isRequired,
  deviceGroupNames: PropTypes.object.isRequired,
  onCheck: PropTypes.func.isRequired,
};
