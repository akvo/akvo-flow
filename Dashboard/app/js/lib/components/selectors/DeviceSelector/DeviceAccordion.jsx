import React from 'react';
import PropTypes from 'prop-types';
// eslint-disable-next-line import/no-unresolved
import Checkbox from 'akvo-flow/components/reusable/Checkbox';

export default class DeviceAccordion extends React.Component {
  state = {
    isAccordionOpen: this.props.deviceGroupIsActive || true,
  };

  onAccordionClick = () => {
    const { isAccordionOpen } = this.state;
    this.setState({ isAccordionOpen: !isAccordionOpen });
  };

  getStyleProps = () => {
    const { isAccordionOpen } = this.state;
    const accordionClass = `accordion ${isAccordionOpen ? 'active' : ''}`;
    const fontClass = `fa fa-chevron-${isAccordionOpen ? 'up' : 'down'}`;
    const panelStyle = isAccordionOpen
      ? { display: 'block' }
      : { display: 'none' };

    return {
      accordionClass,
      fontClass,
      panelStyle,
    };
  };

  selectAllDevice = (_, checked) => {
    const { devices, handleSelectAllDevices } = this.props;
    const selectedDevices = [];

    for (let index = 0; index < devices.length; index++) {
      const device = devices[index];
      selectedDevices.push(device.id);
    }

    handleSelectAllDevices(selectedDevices, checked);
  };

  allDevicesSelected = () => {
    const { selectedDevices, devices } = this.props;
    let allSelected = true;

    // loop through and check if devices is selected
    for (let i = 0; i < devices.length; i++) {
      const device = devices[i];

      if (!selectedDevices.includes(device.id)) {
        allSelected = false;
      }
    }

    return allSelected;
  };

  render() {
    const {
      name,
      devices,
      handleSelectDevice,
      selectedDevices,
      id,
    } = this.props;
    const { accordionClass, fontClass, panelStyle } = this.getStyleProps();

    return (
      <div className="accordion-container">
        <div className={accordionClass}>
          <Checkbox
            id={id}
            name={id}
            checked={this.allDevicesSelected()}
            onChange={this.selectAllDevice}
            label=""
          />

          <div
            onClick={this.onAccordionClick}
            onKeyPress={this.onAccordionClick}
            data-testid="accordion"
          >
            <span>{name}</span>

            <span className={fontClass} />
          </div>
        </div>

        <div className="panel" style={panelStyle} data-testid="panel">
          {devices.map(device => (
            <div key={device.id}>
              <Checkbox
                id={device.id}
                name={device.id}
                checked={selectedDevices.includes(device.id)}
                onChange={handleSelectDevice}
                label={device.name}
              />
            </div>
          ))}
        </div>
      </div>
    );
  }
}

DeviceAccordion.defaultProps = {
  deviceGroupIsActive: true,
};

DeviceAccordion.propTypes = {
  id: PropTypes.string.isRequired,
  deviceGroupIsActive: PropTypes.bool,
  name: PropTypes.string.isRequired,
  devices: PropTypes.array.isRequired,
  handleSelectDevice: PropTypes.func.isRequired,
  handleSelectAllDevices: PropTypes.func.isRequired,
  selectedDevices: PropTypes.array.isRequired,
};
