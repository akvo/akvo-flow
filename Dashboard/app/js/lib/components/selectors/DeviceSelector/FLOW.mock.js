// mock globals variables
const FLOW = {
  deviceGroupControl: {
    content: {
      isLoaded: true,
    },
    get: jest.fn().mockReturnValue({
      forEach: jest.fn(cb =>
        cb({
          get: jest.fn((key) => {
            switch (key) {
            case 'keyId':
              return 1;
            case 'code':
              return 'Devices not in a group';
            default:
              return null;
            }
          }),
        })),
    }),
  },

  deviceControl: {
    content: {
      isLoaded: true,
    },
    get: jest.fn().mockReturnValue({
      forEach: jest.fn(cb =>
        cb({
          get: jest.fn((key) => {
            switch (key) {
            case 'keyId':
              return 150452032;
            case 'deviceGroup':
              return 1;
            case 'deviceIdentifier':
              return 'jana';
            default:
              return null;
            }
          }),
        })),
    }),
  },

  selectedControl: {
    selectedSurveyAssignment: {
      get: jest.fn((key) => {
        switch (key) {
        case 'deviceIds':
          return [150452032];
        default:
          return null;
        }
      }),
    },

    selectedDevices: {
      pushObject: jest.fn(),
      removeObject: jest.fn(),
    },
  },

  Device: {
    find: jest.fn().mockReturnValue({
      get: jest.fn((key) => {
        switch (key) {
        case 'keyId':
          return 150452032;
        default:
          return null;
        }
      }),
    }),
  },
};

global.FLOW = FLOW;

export default { FLOW };
