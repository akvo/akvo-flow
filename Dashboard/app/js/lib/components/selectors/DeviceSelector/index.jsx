import React from 'react';
import PropTypes from 'prop-types';
import DeviceAccordion from './DeviceAccordion';

export default class DeviceSelector extends React.Component {
  render() {
    const {
      deviceGroups,
      handleSelectDevice,
      selectedDevices,
      handleSelectAllDevices,
    } = this.props;

    return (
      <div>
        {Object.keys(deviceGroups).map(dgId => (
          <DeviceAccordion
            key={dgId}
            id={dgId}
            name={deviceGroups[dgId][0].deviceGroup.name}
            devices={deviceGroups[dgId]}
            handleSelectDevice={handleSelectDevice}
            handleSelectAllDevices={handleSelectAllDevices}
            selectedDevices={selectedDevices}
          />
        ))}
      </div>
    );
  }
}

DeviceSelector.propTypes = {
  deviceGroups: PropTypes.object.isRequired,
  handleSelectDevice: PropTypes.func.isRequired,
  handleSelectAllDevices: PropTypes.func.isRequired,
  selectedDevices: PropTypes.array.isRequired,
};
