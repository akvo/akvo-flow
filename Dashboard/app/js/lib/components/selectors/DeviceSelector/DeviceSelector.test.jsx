import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import DeviceSelector from './index';
import '@testing-library/jest-dom/extend-expect';
import flowMock from './FLOW.mock';

// https://github.com/testing-library/react-testing-library#suppressing-unnecessary-warnings-on-react-dom-168
// eslint-disable-next-line no-console
const originalError = console.error;
beforeAll(() => {
  // eslint-disable-next-line no-console
  console.error = (...args) => {
    if (/supports * the "act"/.test(args[0])) {
      return;
    }
    originalError.call(console, ...args);
  };
});

afterAll(() => {
  // eslint-disable-next-line no-console
  console.error = originalError;
});


// Main Tests
describe('DeviceSelector Tests', () => {
  afterEach(cleanup);

  it('+++ renders <snapshot>', () => {
    const wrapper = render(
      <DeviceSelector />
    );

    expect(wrapper.container).toMatchSnapshot();
  });

  it('+++ toggles accordion', () => {
    const wrapper = render(
      <DeviceSelector />
    );

    // find and click on accordion
    const accordion = wrapper.getByTestId('accordion');
    fireEvent.click(accordion);

    // expect panel display to be block
    const panel = wrapper.getByTestId('panel');
    expect(panel).toHaveStyle('display: none'); // open by default when one item was checked
    expect(wrapper.container).toMatchSnapshot();
  });

  test('+++ checkbox works', () => {
    const wrapper = render(
      <DeviceSelector />
    );

    // find input and trigger a check
    // checked by default
    const inputNode = wrapper.getByLabelText('jana');
    fireEvent.click(inputNode);

    expect(flowMock.FLOW.selectedControl.selectedDevices.removeObject).toHaveBeenCalledTimes(1);
  });
});
