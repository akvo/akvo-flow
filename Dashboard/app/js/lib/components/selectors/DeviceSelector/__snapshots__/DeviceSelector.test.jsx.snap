// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceSelector Tests +++ renders <snapshot> 1`] = `
<div>
  <div>
    <div>
      <div
        class="accordion active"
        data-testid="accordion"
      >
        Devices not in a group
      </div>
      <div
        class="panel"
        data-testid="panel"
        style="display: block;"
      >
        <div>
          <input
            checked=""
            class="displayInline"
            id="150452032"
            name="150452032"
            type="checkbox"
          />
          <label
            for="150452032"
            id="150452032"
          >
            jana
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DeviceSelector Tests +++ toggles accordion 1`] = `
<div>
  <div>
    <div>
      <div
        class="accordion false"
        data-testid="accordion"
      >
        Devices not in a group
      </div>
      <div
        class="panel"
        data-testid="panel"
        style="display: none;"
      >
        <div>
          <input
            checked=""
            class="displayInline"
            id="150452032"
            name="150452032"
            type="checkbox"
          />
          <label
            for="150452032"
            id="150452032"
          >
            jana
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
